From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Reilly Brogan <reilly@reillybrogan.com>
Date: Tue, 8 Oct 2024 12:03:06 -0500
Subject: [PATCH] Build system fixes

- Respect CC
- Respect CFLAGS/LDFLAGS
- Disable Werror
- Remove options incompatible with Clang
---
 buildSrc/linux.gradle                          | 18 +++++++++++-------
 .../gstreamer/projects/linux/avplugin/Makefile | 17 ++++++-----------
 .../projects/linux/fxplugins/Makefile          | 18 ++++++------------
 .../projects/linux/gstreamer-lite/Makefile     | 17 ++++++-----------
 .../native/jfxmedia/projects/linux/Makefile    | 17 ++++++-----------
 5 files changed, 35 insertions(+), 52 deletions(-)

diff --git a/buildSrc/linux.gradle b/buildSrc/linux.gradle
index 78571ed2d..c1003a517 100644
--- a/buildSrc/linux.gradle
+++ b/buildSrc/linux.gradle
@@ -43,7 +43,10 @@ LINUX.library = { name -> return (IS_STATIC_BUILD ? "lib${name}.a" : "lib${name}
 def commonFlags = [
         "-fno-strict-aliasing", "-fPIC", "-fno-omit-frame-pointer", // optimization flags
         "-fstack-protector",
-        "-Wextra", "-Wall", "-Wformat-security", "-Wno-unused", "-Wno-parentheses", "-Werror=trampolines"] // warning flags
+        "-Wno-error=cast-function-type",
+        "-Wextra", "-Wall", "-Wformat-security", "-Wno-unused", "-Wno-parentheses"] // warning flags
+
+commonFlags.addAll(System.getenv("CFLAGS").trim().replaceAll(" +", " ").split(" "))
 
 if (!IS_64) {
     commonFlags += "-m32"
@@ -70,6 +73,7 @@ def dynamicLinkFlags = ["-static-libgcc", "-static-libstdc++", "-shared", common
 def staticLinkFlags = [].flatten()
 
 def linkFlags = IS_STATIC_BUILD ? staticLinkFlags : dynamicLinkFlags;
+linkFlags.addAll(System.getenv("LDFLAGS").trim().split(" "))
 
 if (IS_DEBUG_NATIVE) {
     linkFlags += "-g"
@@ -177,8 +181,8 @@ setupTools("linux_freetype_tools",
     }
 )
 
-def compiler = IS_COMPILE_PARFAIT ? "parfait-gcc" : "${toolchainDir}gcc";
-def linker = IS_STATIC_BUILD ? "ar" : IS_COMPILE_PARFAIT ? "parfait-g++" : "${toolchainDir}g++";
+def compiler = IS_COMPILE_PARFAIT ? "parfait-gcc" : System.getenv("CC");
+def linker = IS_STATIC_BUILD ? "ar" : IS_COMPILE_PARFAIT ? "parfait-g++" : System.getenv("CXX");
 
 // Strip native .so shared libraries as a postprocess step when copying them
 LINUX.strip = "${toolchainDir}strip"
@@ -198,7 +202,7 @@ FileTree ft_gtk = fileTree("${project(":graphics").projectDir}/src/main/native-g
 LINUX.glass.glass = [:]
 LINUX.glass.glass.nativeSource = ft_gtk_launcher.getFiles()
 LINUX.glass.glass.compiler = compiler
-LINUX.glass.glass.ccFlags = [cppFlags, "-Werror"].flatten()
+LINUX.glass.glass.ccFlags = [cppFlags].flatten()
 LINUX.glass.glass.linker = linker
 LINUX.glass.glass.linkFlags = IS_STATIC_BUILD? linkFlags : [linkFlags, "-lX11", "-ldl"].flatten()
 LINUX.glass.glass.lib = "glass"
@@ -208,7 +212,7 @@ LINUX.glass.glassgtk3.nativeSource =  ft_gtk.getFiles()
 LINUX.glass.glassgtk3.compiler = compiler
 LINUX.glass.glassgtk3.ccFlags = IS_STATIC_BUILD ?
         ["-fno-threadsafe-statics", cppFlags, gtk3CCFlags].flatten() :
-        [cppFlags, gtk3CCFlags, "-Werror"].flatten()
+        [cppFlags, gtk3CCFlags].flatten()
 LINUX.glass.glassgtk3.linker = linker
 LINUX.glass.glassgtk3.linkFlags = IS_STATIC_BUILD ? linkFlags : [linkFlags, gtk3LinkFlags].flatten()
 LINUX.glass.glassgtk3.lib = "glassgtk3"
@@ -243,7 +247,7 @@ LINUX.iio.nativeSource = [
 LINUX.iio.compiler = compiler
 LINUX.iio.ccFlags = [cFlags, "-fvisibility=hidden"].flatten()
 LINUX.iio.linker = IS_STATIC_BUILD ? "ld" : linker
-LINUX.iio.linkFlags = [linkFlags].flatten()
+LINUX.iio.linkFlags = [linkFlags, "-ljpeg"].flatten()
 LINUX.iio.lib = "javafx_iio"
 
 LINUX.prismES2 = [:]
@@ -291,5 +295,5 @@ LINUX.media.ar = "${toolchainDir}ar"
 LINUX.webkit = [:]
 LINUX.webkit.compiler = compiler
 LINUX.webkit.linker = linker
-LINUX.webkit.ccFlags = commonFlags.flatten()
+LINUX.webkit.ccFlags = [commonFlags, "-fno-lto"].flatten()
 LINUX.webkit.linkFlags = linkFlags.flatten()
diff --git a/modules/javafx.media/src/main/native/gstreamer/projects/linux/avplugin/Makefile b/modules/javafx.media/src/main/native/gstreamer/projects/linux/avplugin/Makefile
index 4081fbe97..15178e3c9 100644
--- a/modules/javafx.media/src/main/native/gstreamer/projects/linux/avplugin/Makefile
+++ b/modules/javafx.media/src/main/native/gstreamer/projects/linux/avplugin/Makefile
@@ -17,13 +17,12 @@ TARGET = $(BUILD_DIR)/lib$(BASE_NAME).so
 OBJBASE_DIR = $(BUILD_DIR)/obj/plugins/avplugin
 endif
 
-CFLAGS = -fPIC                   \
+CFLAGS := -fPIC                   \
          -Wformat                \
          -Wextra                 \
          -Wformat-security       \
          -fstack-protector       \
          -Werror=implicit-function-declaration \
-         -Werror=trampolines     \
          -fbuiltin               \
          -DHAVE_STDINT_H         \
          -DLINUX                 \
@@ -31,18 +30,13 @@ CFLAGS = -fPIC                   \
          -DGSTREAMER_LITE \
          -DGLIB_VERSION_MIN_REQUIRED=GLIB_VERSION_2_48 \
          -DGLIB_VERSION_MAX_ALLOWED=GLIB_VERSION_2_48 \
-         -ffunction-sections -fdata-sections
+         -ffunction-sections -fdata-sections \
+         ${CFLAGS}
 
 ifneq (,$(findstring $(ARCH), x64 x32))
     CFLAGS += -msse2
 endif
 
-ifeq ($(BUILD_TYPE), Release)
-    CFLAGS += -Os
-else
-    CFLAGS += -g -Wall
-endif
-
 PACKAGES_INCLUDES := $(shell pkg-config --cflags glib-2.0)
 PACKAGES_LIBS     := $(shell pkg-config --libs glib-2.0 gobject-2.0)
 
@@ -52,11 +46,12 @@ INCLUDES= -I../../../plugins                          \
           -I../../../gstreamer-lite/gstreamer/libs    \
           $(PACKAGES_INCLUDES)
 
-LDFLAGS = -L$(BUILD_DIR)    \
+LDFLAGS := -L$(BUILD_DIR)    \
           -lgstreamer-lite  \
            $(PACKAGES_LIBS) \
           -z relro          \
-          -static-libgcc -static-libstdc++ -Wl,--gc-sections
+          -static-libgcc -static-libstdc++ -Wl,--gc-sections \
+         ${LDFLAGS}
 
 ifneq ($(strip $(LIBAV_DIR)),)
 INCLUDES += -I$(LIBAV_DIR)/include
diff --git a/modules/javafx.media/src/main/native/gstreamer/projects/linux/fxplugins/Makefile b/modules/javafx.media/src/main/native/gstreamer/projects/linux/fxplugins/Makefile
index 0470a364c..89493d662 100644
--- a/modules/javafx.media/src/main/native/gstreamer/projects/linux/fxplugins/Makefile
+++ b/modules/javafx.media/src/main/native/gstreamer/projects/linux/fxplugins/Makefile
@@ -14,13 +14,12 @@ DIRLIST = progressbuffer       \
 
 TARGET = $(BUILD_DIR)/lib$(BASE_NAME).so
 
-CFLAGS = -fPIC                   \
+CFLAGS := -fPIC                   \
          -Wformat                \
          -Wextra                 \
          -Wformat-security       \
          -fstack-protector       \
          -Werror=implicit-function-declaration \
-         -Werror=trampolines     \
          -Werror=deprecated-declarations \
          -fbuiltin               \
          -DHAVE_STDINT_H         \
@@ -33,19 +32,13 @@ CFLAGS = -fPIC                   \
          -DGSTREAMER_LITE \
          -DGLIB_VERSION_MIN_REQUIRED=GLIB_VERSION_2_48 \
          -DGLIB_VERSION_MAX_ALLOWED=GLIB_VERSION_2_48 \
-         -ffunction-sections -fdata-sections
+         -ffunction-sections -fdata-sections \
+         ${CFLAGS}
 
 ifneq (,$(findstring $(ARCH), x64 x32))
     CFLAGS += -msse2
 endif
 
-ifeq ($(BUILD_TYPE), Release)
-    CFLAGS += -Os
-else
-    CFLAGS += -g -Wall
-endif
-
-
 INCLUDES = -I$(SRCBASE_DIR) \
 	   $(addprefix -I$(SRCBASE_DIR)/,$(DIRLIST)) \
            -I../../../gstreamer-lite/gstreamer       \
@@ -54,9 +47,10 @@ INCLUDES = -I$(SRCBASE_DIR) \
 PACKAGES_INCLUDES := $(shell pkg-config --cflags glib-2.0)
 PACKAGES_LIBS := $(shell pkg-config --libs glib-2.0 gobject-2.0)
 
-LDFLAGS = -L$(BUILD_DIR) -lgstreamer-lite $(PACKAGES_LIBS) \
+LDFLAGS := -L$(BUILD_DIR) -lgstreamer-lite $(PACKAGES_LIBS) \
           -z relro \
-          -static-libgcc -static-libstdc++ -Wl,--gc-sections
+          -static-libgcc -static-libstdc++ -Wl,--gc-sections \
+          ${LDFLAGS}
 
 ifeq ($(ARCH), x32)
     CFLAGS += -m32
diff --git a/modules/javafx.media/src/main/native/gstreamer/projects/linux/gstreamer-lite/Makefile b/modules/javafx.media/src/main/native/gstreamer/projects/linux/gstreamer-lite/Makefile
index 374f50982..ccd7ff178 100644
--- a/modules/javafx.media/src/main/native/gstreamer/projects/linux/gstreamer-lite/Makefile
+++ b/modules/javafx.media/src/main/native/gstreamer/projects/linux/gstreamer-lite/Makefile
@@ -37,13 +37,12 @@ DIRLIST = gstreamer/gst \
 
 TARGET = $(BUILD_DIR)/lib$(BASE_NAME).so
 
-CFLAGS =-fPIC                   \
+CFLAGS := -fPIC                   \
         -Wformat                \
         -Wextra                 \
         -Wformat-security       \
         -fstack-protector       \
         -Werror=implicit-function-declaration \
-        -Werror=trampolines     \
         -Werror=deprecated-declarations \
         -D_GNU_SOURCE           \
         -DGST_REMOVE_DEPRECATED \
@@ -55,13 +54,8 @@ CFLAGS =-fPIC                   \
         -DGST_DISABLE_LOADSAVE  \
         -DGLIB_VERSION_MIN_REQUIRED=GLIB_VERSION_2_48 \
         -DGLIB_VERSION_MAX_ALLOWED=GLIB_VERSION_2_48 \
-        -ffunction-sections -fdata-sections
-
-ifeq ($(BUILD_TYPE), Release)
-    CFLAGS += -Os
-else
-    CFLAGS += -g -Wall
-endif
+        -ffunction-sections -fdata-sections \
+        ${CFLAGS}
 
 INCLUDES = -I$(BASE_DIR)/plugins                          \
            -I$(SRCBASE_DIR)/projects/build/linux/common   \
@@ -79,9 +73,10 @@ INCLUDES = -I$(BASE_DIR)/plugins                          \
 PACKAGES_INCLUDES := $(shell pkg-config --cflags alsa glib-2.0)
 PACKAGES_LIBS := $(shell pkg-config --libs alsa glib-2.0 gobject-2.0 gmodule-2.0 gthread-2.0)
 
-LDFLAGS = -L$(BUILD_DIR) -lm $(PACKAGES_LIBS) \
+LDFLAGS := -L$(BUILD_DIR) -lm $(PACKAGES_LIBS) \
           -z relro \
-          -static-libgcc -static-libstdc++ -Wl,--gc-sections
+          -static-libgcc -static-libstdc++ -Wl,--gc-sections \
+          ${LDFLAGS}
 
 ifeq ($(ARCH), x32)
     CFLAGS += -m32
diff --git a/modules/javafx.media/src/main/native/jfxmedia/projects/linux/Makefile b/modules/javafx.media/src/main/native/jfxmedia/projects/linux/Makefile
index 6a1ad3b9b..a555854c0 100644
--- a/modules/javafx.media/src/main/native/jfxmedia/projects/linux/Makefile
+++ b/modules/javafx.media/src/main/native/jfxmedia/projects/linux/Makefile
@@ -42,7 +42,7 @@ DIRLIST = jni                \
 
 TARGET = $(BUILD_DIR)/lib$(BASE_NAME).so
 
-CFLAGS = -DTARGET_OS_LINUX=1     \
+CFLAGS := -DTARGET_OS_LINUX=1     \
          -D_GNU_SOURCE           \
          -DGST_REMOVE_DEPRECATED \
          -DGST_DISABLE_GST_DEBUG \
@@ -54,7 +54,8 @@ CFLAGS = -DTARGET_OS_LINUX=1     \
          -DGLIB_VERSION_MIN_REQUIRED=GLIB_VERSION_2_48 \
          -DGLIB_VERSION_MAX_ALLOWED=GLIB_VERSION_2_48 \
          -Werror=deprecated-declarations \
-         -ffunction-sections -fdata-sections
+         -ffunction-sections -fdata-sections \
+         ${CFLAGS}
 
 CPPFLAGS = -fno-rtti -ffunction-sections -fdata-sections
 
@@ -70,7 +71,6 @@ ifdef HOST_COMPILE
                   -Wextra \
                   -Wformat-security \
                   -fstack-protector \
-                  -Werror=trampolines \
 	          -DGSTREAMER_LITE
         ifneq (,$(findstring $(ARCH), x64 x32))
             CFLAGS += -msse2
@@ -89,21 +89,16 @@ ifdef HOST_COMPILE
 	           -I$(PLUGINS_DIR) \
 		  $(PACKAGES_INCLUDES)
 
-	LDFLAGS = -Wl,-rpath,\$$ORIGIN -L$(BUILD_DIR) -lgstreamer-lite $(PACKAGES_LIBS) \
+	LDFLAGS := -Wl,-rpath,\$$ORIGIN -L$(BUILD_DIR) -lgstreamer-lite $(PACKAGES_LIBS) \
             -z relro \
-            -static-libgcc -static-libstdc++ -Wl,--gc-sections
+            -static-libgcc -static-libstdc++ -Wl,--gc-sections \
+            ${LDFLAGS}
 else
 	CFLAGS += $(EXTRA_CFLAGS)
 	INCLUDES = $(BASE_INCLUDES)
 	LDFLAGS = -Wl,-rpath,\$$ORIGIN -L$(BUILD_DIR) $(EXTRA_LDFLAGS)
 endif
 
-ifeq ($(BUILD_TYPE), Release)
-	CFLAGS += -Os
-else
-	CFLAGS += -g -Wall -D_DEBUG
-endif
-
 ifeq ($(ARCH), x32)
     CFLAGS += -m32
     LDFLAGS += -m32
